assembly/src/assembler.rs

```diff
@@ -110,7 +110,10 @@ impl Assembler {
     }
 
     /// Start building an [`Assembler`] with a kernel defined by the provided [KernelLibrary].
-    pub fn with_kernel(source_manager: Arc<dyn SourceManager>, kernel_lib: KernelLibrary) -> Self {
+    pub fn with_kernel(
+        source_manager: Arc<dyn SourceManagerSync>,
+        kernel_lib: KernelLibrary,
+    ) -> Self {
```

Comment:
I think we should probably revert this change so that the assembler does not make assumptions about whether we are using sync or non-sync source manager.